main.py

```diff
@@ -112,55 +131,68 @@ def generate_audio(self, phonemes: str, voice: str, speed: float) -> np.ndarray:
-            if level < self.SILENCE_THRESHOLD:
-                silence_frames += len(audio)
-            else:
-                silence_frames = 0
-
-            # process audio when silence is detected
-            if silence_frames > self.SILENCE_DURATION * self.SAMPLE_RATE:
-                audio_segment = np.array(audio_buffer, dtype=np.float32)
-
-                if len(audio_segment) > self.SAMPLE_RATE:
-                    text = self.whisper_mlx.transcribe(audio_segment)['text']
-
-                    # skip empty/invalid transcriptions
-                    if text.strip():
-                        print(f"Transcription: {text}")
-                        self.messages.append({
-                            'role': 'user',
-                            'content': text
-                        })
-                        self.create_and_play_response(text)
-
-                # reset state
-                audio_buffer.clear()
-                silence_frames = 0
-                total_frames = 0
-
-        # start recording loop
+
+            # Process audio in chunks
+            if len(audio_buffer) >= self.AUDIO_CHUNK_SIZE:
+                audio_chunk = np.array(audio_buffer[:self.AUDIO_CHUNK_SIZE], dtype=np.float32)
+                audio_buffer = audio_buffer[self.AUDIO_CHUNK_SIZE:]  # Remove processed chunk from buffer
+
+                # Resample to 16 kHz (required by Silero VAD)
+                audio_chunk_resampled = resample(audio_chunk, int(self.WHISPER_SAMPLE_RATE * self.AUDIO_CHUNK_DURATION))
+
+                # Check for voice activity
+                speech_timestamps = self.get_speech_timestamps(audio_chunk_resampled, self.vad_model, sampling_rate=self.WHISPER_SAMPLE_RATE)
+
+                if speech_timestamps:
+                    if not is_speaking:
+                        print("Voice detected, starting transcription...")
+                        is_speaking = True
+
+                    # Accumulate audio during speech detection
+                    accumulated_audio.extend(audio_chunk_resampled.tolist())
+                else:
+                    if is_speaking:
+                        print("Silence detected, transcribing...")
+                        is_speaking = False
+
+                        # Transcribe accumulated audio
+                        if accumulated_audio:
+                            accumulated_audio_np = np.array(accumulated_audio, dtype=np.float32)
+                            text = self.whisper_mlx.transcribe(accumulated_audio_np)['text']
+
+                            # Validate transcription
+                            if text.strip():
+                                if text.strip() in BAD_SENTENCES:
+                                    print('Hallucination detected. Skipping.')
+                                elif not re.fullmatch(r"[A-Za-z\s.,!?\'\"-:;()]*", text):
+                                    print('Non-english characters detected. Skipping.', text)
```

Comment:
I found whisper to transcribe weird characters, this check is to ignore that. However, this does restrict it to english only, so feel free to take this out.